pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,353 @@
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+			for _, oidBlock := range blockOids {
+				go func(oidBlock oid.ID) {
+					data, err := bfs.get(oidBlock.String())
+					if err != nil {
+						bfs.log.Error(fmt.Sprintf("Failed to fetch block %s", oidBlock.String()), zap.Error(err))
+					}
+					err = bfs.processBlock(data)
+					if err != nil {
+						bfs.log.Error(fmt.Sprintf("Failed to process block with index %d", startIndex), zap.Error(err))
+					}
+				}(oidBlock)
+			}
+
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchDataWithOid() error {
+	startIndex := bfs.chain.BlockHeight() / 2000
+	skip := bfs.chain.BlockHeight() % 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter("block_oids", fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter("block_oids", fmt.Sprintf("%d", startIndex), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			blockOidsObject, err := bfs.search(prm)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("No 'block_oids' object found with index %d, stopping.", startIndex))
+				return errors.New("no 'block_oids' object found")
+			}
+
+			blockOidsData, err := bfs.get(blockOidsObject[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'block_oids' object %d", startIndex), zap.Error(err))
+				return err
+			}
+			blockOIDs, err := parseBlockOIDs(blockOidsData)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to parse 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			for j, oidBlock := range blockOIDs {
+				if j < int(skip) {
+					continue
+				}
+				data, err := bfs.get(oidBlock.String())
+				if err != nil {
+					bfs.log.Error(fmt.Sprintf("Failed to fetch block with OID %s", oidBlock.String()), zap.Error(err))
+					return err
+				}
+				bfs.log.Info(fmt.Sprintf("Fetched block with OID %s", oidBlock.String()))
+				err = bfs.processBlock(data)
+				if err != nil {
+					bfs.log.Error(fmt.Sprintf("Failed to process blocks with index %d", startIndex), zap.Error(err))
+					return err
+				}
+			}
+
+			startIndex++
+		}
+	}
+}
+
+func (bfs *Service) processBlock(data []byte) error {
+	var buf []byte
+	stateRootInHeader := bfs.chain.GetConfig().StateRootInHeader
+	br := gio.NewBinReaderFromBuf(data)
+
+	readBlock := func(r *gio.BinReader) ([]byte, error) {
+		var size = r.ReadU32LE()
+		if uint32(cap(buf)) < size {
+			buf = make([]byte, size)
+		} else {
+			buf = buf[:size]
+		}
+		r.ReadBytes(buf)
+		return buf, r.Err
```

Comment:
You just don't need these size-related manipulations if `data` contains only block bytes. Just use `(b *Block).DecodeBinary`.